### PR TITLE
Geometry dimensions via HasDimensions trait

### PR DESCRIPTION
**Pull request by michaelkirk**
Some efforts extracted from the geomgraph work that seemed generally useful, and thus worth separating.

(I'm also trying to do what I can to keep the upcoming geomgraph PR just a little bit smaller.)

*Note this PR represents separate functionality from #523, but it relies on #523 for correct behavior, so I'm targeting #523 as the base branch. Before merging we should update the base to target master.*

### CODE REVIEW

File: geo/src/algorithm/dimensions.rs [hunk 1]
```diff
@@ -0,0 +1,452 @@
...
+/// let rect = Rect::new((0.0, 0.0), (10.0, 10.0));
+/// assert_eq!(Dimensions::ZeroDimensional, point.get_dimensions());
+/// assert_eq!(Dimensions::OneDimensional, line_string.get_dimensions());
+/// assert_eq!(Dimensions::TwoDimensional, rect.get_dimensions());
+///
+/// assert!(point.get_dimensions() < line_string.get_dimensions());
+/// assert!(rect.get_dimensions() > line_string.get_dimensions());
+/// ```
+#[derive(Debug, Clone, Copy, PartialEq, Eq, Ord, PartialOrd)]
+pub enum Dimensions {
```
Comment (michaelkirk) [inline geo/src/algorithm/dimensions.rs:26]: "Dimensionality" would probably be a more "correct" term for this enum, but I thought maybe it was too verbose. 

I could be persuaded!

File: geo/src/algorithm/dimensions.rs [hunk 2]
```diff
@@ -0,0 +1,452 @@
...
+    /// Some geometries, like a `MultiPoint` or `GeometryColletion` may have no elements - thus no
+    /// dimensions. Note that this is distinct from being `ZeroDimensional`, like a `Point`.
+    Empty,
```
Comment (michaelkirk) [inline geo/src/algorithm/dimensions.rs:29]: I considered ommitting this and instead representing Empty as a `None` `Option<Dimensions>`, but it seemed to work better with `Empty` as an explicit member of the enum. 

Primarily because I want to use it in comparisons:

```
if a.get_dimensions() < Dimensions::OneDimensional {
    do_something()
} else {
    do_something_else()
}
```

is preferable to something like:

```
match a.get_dimensions() {
    Some(dimension) => {
        if dimensions < Dimensions::OneDimensional {
            do_something()
        } else {
            do_something_else()
        } 
    }
    None => {  do_something_else() }
}
```

File: geo/src/algorithm/dimensions.rs (hunk 1 above)
Comment (urschrei) [reply geo/src/algorithm/dimensions.rs:26]: It's our code, and we do what we feel like. (There's no ambiguity in "dimensions", so there's no problem imo)

File: geo/src/algorithm/dimensions.rs [hunk 3]
```diff
@@ -0,0 +1,452 @@
...
+    ///
+    /// let point = Point::new(10.0, 10.0);
+    /// assert_eq!(Dimensions::ZeroDimensional, point.get_dimensions());
+    ///
+    /// // An `Empty` dimensionality is distinct from, and less than, being 0-dimensional
+    /// let empty_collection = GeometryCollection::<f32>(vec![]);
+    /// assert_eq!(Dimensions::Empty, empty_collection.get_dimensions());
+    /// assert!(empty_collection.get_dimensions() < point.get_dimensions());
+    /// ```
+    fn get_dimensions(&self) -> Dimensions;
```
Comment (frewsxcv) [inline geo/src/algorithm/dimensions.rs:97]: thoughts on removing the `get_` prefix? getters in rust tend to not have 'get'

https://rust-lang.github.io/api-guidelines/naming.html#getter-names-follow-rust-convention-c-getter

File: geo/src/algorithm/dimensions.rs [hunk 4]
```diff
@@ -0,0 +1,452 @@
...
+        let first = self.0[0];
+        if self.0.iter().any(|&coord| first != coord) {
+            Dimensions::OneDimensional
+        } else {
+            // all coords are the same - i.e. a point
+            Dimensions::ZeroDimensional
+        }
+    }
+
+    ///```
```
Comment (frewsxcv) [inline geo/src/algorithm/dimensions.rs:239]:
```suggestion
    /// ```
```

**Review by frewsxcv — APPROVED** [review]
🎉

**Review by urschrei — APPROVED** [review]
This PR is a _delight_

File: geo/src/algorithm/dimensions.rs (hunk 3 above)
Comment (michaelkirk) [reply geo/src/algorithm/dimensions.rs]: Love it! Honestly I assumed you couldn't have a method and property with the same name. 😅

File: geo/src/algorithm/dimensions.rs [hunk 5]
```diff
@@ -0,0 +1,452 @@
...
+    }
+
+    fn boundary_dimensions(&self) -> Dimensions {
+        Dimensions::Empty
+    }
+}
+
+impl<C: CoordinateType> HasDimensions for MultiLineString<C> {
+    fn is_empty(&self) -> bool {
+        if self.0.is_empty() {
```
Comment (rmanoka) [inline geo/src/algorithm/dimensions.rs:296]: Why is this needed?  Iterator::all returns true on empty iterator (Iterator::any returns false).

Comment (michaelkirk) [reply geo/src/algorithm/dimensions.rs]: Good point - removed!

**Comment by michaelkirk** [comment]
I think I've addressed all your feedback here as well @rmanoka. PTAL

**Comment by michaelkirk** [comment]
@rmanoka - I believe I've addressed your feedback. Unless I hear that you'd like more time to review, I'll merge in 24 hours.

**Review by rmanoka — APPROVED** [review]

**Comment by rmanoka** [comment]
@michaelkirk lgtm, please go ahead.

**Comment by michaelkirk** [comment]
bors r=rmanoka
